crates/m3/tests/merkle_tree.rs

```diff
@@ -182,96 +235,152 @@ mod model {
 	}
 	impl MerkleTreeTrace {
 		/// Method to generate the trace given the witness values. The function assumes that the
 		/// root_id is the index of the root in the roots vector and that the paths and leaves are
 		/// passed in with their assigned root_id.
 		fn generate(roots: Vec<[u8; 32]>, paths: &[MerklePath]) -> Self {
-			let mut path_vec = Vec::new();
-			let mut root_vec = Vec::new();
-			let mut boundary_vec: Vec<MerkleBoundary> = Vec::new();
-			// Number of times each root is referenced in the paths.
-			let mut root_multiplicities = vec![0; roots.len()];
-
-			for (root_id, index, leaf, path) in paths.iter() {
-				// Push the boundary values for the statement.
-				boundary_vec.push(MerkleBoundary {
-					leaf: (*root_id, *leaf, path.len(), *index),
-					root: (*root_id, roots[*root_id as usize]),
-				});
-
-				root_multiplicities[*root_id as usize] += 1;
-
-				let mut leaf = *leaf;
-				for (i, node) in path.iter().enumerate() {
-					let mut parent = [0u8; 32];
-					if (index >> i) & 1 == 0 {
-						compress(&leaf, node, &mut parent);
-						path_vec.push(MerklePathEvent {
-							root_id: *root_id,
-							left: leaf,
-							right: *node,
-							parent,
-							parent_depth: path.len() - i - 1,
-							parent_index: index >> (i + 1),
-							flush_left: true,
-							flush_right: false,
-						});
-					} else {
-						compress(node, &leaf, &mut parent);
-						path_vec.push(MerklePathEvent {
-							root_id: *root_id,
-							left: *node,
-							right: leaf,
-							parent,
-							parent_depth: path.len() - i - 1,
-							parent_index: index >> (i + 1),
-							flush_left: false,
-							flush_right: true,
-						});
+			let mut path_nodes = Vec::new();
+			let mut root_nodes = HashSet::new();
+			let mut boundaries = MerkleBoundaries::new();
+			// Number of times each root is referenced in the paths. Since internal nodes have been
+			// deduped, these need to be pushed into the nodes channel as many times as they are
+			// referenced in the paths.
+
+			//tracks the filled nodes in the tree
+			let mut filled_nodes = HashMap::new();
+
+			for path in paths.iter() {
+				let MerklePath {
+					root_id,
+					index,
+					leaf,
+					nodes,
+				} = path;
+
+				let mut current_child = *leaf;
+
+				// Populate the boundary values for the statement.
+				boundaries.insert(
+					NodeFlushToken {
+						root_id: *root_id,
+						data: current_child,
+						depth: nodes.len(),
+						index: *index,
+					},
+					RootFlushToken {
+						root_id: *root_id,
+						data: roots[*root_id as usize],
+					},
+				);
+
+				root_nodes.insert(MerkleRootEvent::new(*root_id, roots[*root_id as usize]));
+
+				let mut parent_node = [0u8; 32];
+				for (i, &node) in nodes.iter().enumerate() {
+					match filled_nodes.get_mut(&(*root_id, index >> (i + 1), nodes.len() - i - 1)) {
+						Some((_, _, _, flush_left, flush_right)) => {
+							if (index >> i) & 1 == 0 {
+								compress(&current_child, &node, &mut parent_node);
```

Comment:
You shouldn't need to recompress, you can read `parent_node` out of the `filled_nodes` value.